### DISCUSSION TITLE
Skeletal Animation

### DISCUSSION THREAD

**Discussion by voxelias** [bevyengine/bevy#508] [Ideas]
Hello World! I am working on skeletal animation implementation for Bevy. My first goal is to have working example demonstrating one animation loaded from a GLTF file. 

The plan is:

- [x] Extend Mesh asset with `joints_indices` (up to 4 indices of joints affecting a point in 3D space) and `joints_weights` (up to 4 weights affecting a point in 3D space) attributes, that shall be passed to the shader. 
- [x] Extend GLTF loader with joints and weights loading
- [ ] Prepare an Animation Assset that will contain keyframes and transformations for joints
- [ ] Prepare a shader, that will take   `joints_indices` and `joints_weights` from VAO and transformations from the uniform
- [ ] Prepare an animation system that will calculate approximate transformations between two keyframes and pass them as a uniform to the shader
- [ ] Prepare an animation loader. I want to do it inside of a startup system for now, because the question about extending/rewriting of the assets system is still open
- [ ] Bring all together in a new example code

Please, let me know what you think? I am new to Bevy and just started with the engine research and may doing something from the wrong end.

**Comment by StarArawn** [comment]
This sounds really good! I think eventually we'll want to use a compute shader to calculate animations, but currently it's not practical to use compute in bevy.

**Reply by zicklag** [reply]
Just curious, in a short, high-level description, what makes it impractical at the moment?

**Reply by StarArawn** [reply]
Bevy doesn't support compute shaders yet.

**Comment by cart** [comment]
I agree that your points all sound good! I think in the immediate short term it makes sense to build the animation part specifically for skeletal animation, but at some point i think it would be interesting to build it into bevy's eventual "unified animation system".

**Comment by TheButlah** [comment]
This update from the 0.6 blog post made me very happy :)
> Skeletal Animation: Ultimately Bevy will have a general purpose, property based animation system (we already have a working implementation). We've been holding off on adding skeletal animation, so we can slot it in to that system, but in retrospect that was a mistake. People need skeletal animation now. In the short term we will build a scoped 3d skeletal animation system, just to get the ball rolling. Then later we will port it to the general purpose system (whenever that is ready),

**Comment by wallabra** [comment]
Ooh! It would be amazing if we could update vertices from the bone positions changing in real time, assuming the influence matrix computations wouldn't be prohibitively expensive (I need to read more about it). If nothing else, this would solve a sizeable dilemma my project faces: https://gamecircular.forumeiros.com/t8-dev-soft-bodies-and-3d-modelling
